lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -201,6 +241,21 @@ static std::string getSubModuleName(Operation *oldOp) {
       subModuleName += "_seq";
   }
 
+  // Add control information.
+  if (auto ctrlAttr = oldOp->getAttrOfType<BoolAttr>("control");
+      ctrlAttr && ctrlAttr.getValue()) {
+    // Add some additional discriminating info for non-typed operations.
+    subModuleName += "_" + std::to_string(oldOp->getNumOperands()) + "ins_" +
+                     std::to_string(oldOp->getNumResults()) + "outs";
+    subModuleName += "_ctrl";
+  } else {
+    if (inTypes.empty() && outTypes.empty()) {
+      oldOp->emitError()
+          << "Non-control operators must provide discriminating type info";
+      assert(false);
```

Comment:
it's a little weird to emit an error and assert. IMO errors are for users of the compiler, and asserts are for engineers of the compiler. if this should be an assertion, i'd suggest to replace this with something like `assert(!(inTypes.empty() && outTypes.empty()) && "non-control operators must provide...")`. or if you want to raise an error, pull this check up higher to a point where you can emit the error and terminate the pass nicely.